libs/graph-lib/src/services/encryption.service.ts

```diff
@@ -0,0 +1,123 @@
+/* eslint-disable max-classes-per-file */
+import { Inject, Injectable, Logger } from '@nestjs/common';
+import { GraphKeyPairDto } from '#types/dtos/graph/graph-key-pair.dto';
+import graphCommonConfig, { IGraphCommonConfig } from '#config/graph-common.config';
+import { Aes256Gcm, CipherSuite, HkdfSha256 } from '@hpke/core';
+import { DhkemX25519HkdfSha256 } from '@hpke/dhkem-x25519';
+import { hexToU8a, u8aToHex } from '@polkadot/util';
+import { HexString } from '@polkadot/util/types';
+import lodash from 'lodash';
+
+export class EncryptionResult {
+  senderContext: HexString;
+
+  encryptionPublicKey: HexString;
+
+  result: GraphKeyPairDto[];
+}
+
+@Injectable()
+export class EncryptionService {
+  private readonly logger: Logger;
+
+  private readonly suite: CipherSuite;
+
+  private encryptionKey: CryptoKeyPair;
+
+  constructor(@Inject(graphCommonConfig.KEY) private readonly config: IGraphCommonConfig) {
+    this.logger = new Logger(this.constructor.name);
+    this.suite = new CipherSuite({
+      kem: new DhkemX25519HkdfSha256(),
+      kdf: new HkdfSha256(),
+      aead: new Aes256Gcm(),
+    });
+  }
+
+  private async init() {
+    if (this.encryptionKey === null || this.encryptionKey === undefined) {
+      this.encryptionKey = await this.suite.kem.deriveKeyPair(
+        new TextEncoder().encode(this.config.atRestEncryptionKeySeed),
+      );
+    }
+  }
+
+  /**
+   * replaces private keys inside GraphKeyPairDto array with an encrypted one
+   * @param graphKeyPairs
+   * @returns the public key hex of encryption key used or returns null if encryption was not required
+   */
+  public async encryptPrivateKeys(graphKeyPairs?: GraphKeyPairDto[]): Promise<EncryptionResult | null> {
+    await this.init();
+
+    if (graphKeyPairs !== null && graphKeyPairs.length > 0) {
+      const sender = await this.suite.createSenderContext({
+        recipientPublicKey: this.encryptionKey.publicKey,
+      });
+
+      const decryptedKeyPairs: GraphKeyPairDto[] = [];
+      for (let i = 0; i < graphKeyPairs.length; i += 1) {
+        const keyPair = lodash.cloneDeep(graphKeyPairs[i]);
```

Comment:
~Taking a dependency on lodash is a hot-button topic... do we need to? It adds a lot of bloat to the dependency tree.~
nvm--looks like NestJS already adds a dependency on lodash ☹️

Reply:
Or better, use native ES6 `structuredClone`:
```suggestion
        const keyPair = structuredClone(graphKeyPairs[i]);
```

Reply:
I was looking for an alternative, thanks Was kinda iffy adding lodash just because of this.